### ISSUE TITLE
Git-based versioning + Tarball code delivery

### ISSUE THREAD

**Issue by TimothyEDawson** [EdwardALuke/loci#35]
As part of the build system refactoring (#20), we will utilize git tags to associate specific commits to each version number increment, generally following the semantic versioning scheme of MAJOR.MINOR.PATCH.

With the version number tracked via tags, the current process of hard-coding the version number within lpp/main.cc and src/System/Loci_version.cc becomes redundant. Under this issue, additional logic will be added to the Make system to extract the version from the git tags using the command `git describe --abbrev=6 --always --dirty --tags`, as well as the branch name with `git rev-parse --abbrev-ref HEAD`. This will be made available to the Loci::version() command alongside the compile date and time to give users more detailed information regarding the source code status at the time of solver compilation.

This version extraction process originally implemented by CFDRC will be reworked in order to support compilation without access to the .git directory, such as when source code is delivered via tarball. Because this sort of code delivery is dependent on how the versioning is handled, this issue will also implement a `make tarball` command to handle generating the version information and packaging up the source code for delivery.

**Comment by EdwardALuke** [comment]
Just to note one of the challenges to the way the current configure system works, we presently can use the /usr/tmp directory to store object files for the compile using the command

./configure --obj-dir=/usr/tmp/<userid>/Loci

git commands executed under the /usr/tmp/<userid>/Loci will fail unless we manage passing information about the location of the git database.

This has been something that I have frequently done in my build process.  I have checked and on the current msstate systems the performance savings from this is not significant (probably due to the very large cache sizes available to the network file systems on modern machines).  But it is still useful for compiling on systems where tight quotas exist on the directories available to users.  This should be included when testing the changes to the build system.

**Comment by TimothyEDawson** [comment]
@EdwardALuke I appreciate that note. One option I had been mulling over would be to have the configure script generate the version information, which would resolve that issue, but I don't think that would be the right thing to do. I think it needs to remain part of the build system to ensure the version info generated stays in sync with the source code being built. One option might be to have the configure script include a symbolic link to the .git directory, I would need to test that thoroughly to make sure it works.

I would also be interested in your thoughts on the process. My current plan is to generate a temporary file separate from main.cc and Loci_version.cc which contains the extracted info, and modify those routines to read from that file into revision_name. When executing `make tarball`, this file would then be generated and included in the packaged delivery, while the git data would be omitted.

Alternatively I could have it generate those files directly. But having a separate file has some advantages:

1. It contains only the modified information, and not the rest of the logic.
2. It won't register as a modified source file every time you compile.
3. You can easily fall back to an "undefined" version number if both that file and the .git folder are not present.

One thing that I still need to figure out is how to ensure it updates the file whenever the version information has changed. Probably the best thing to do is simply to force it to always regenerate the file regardless of whether the version string has changed or not, and I have seen a method to do that.

**Comment by EdwardALuke** [comment]
I agree that we should be running git on every compile so we can get a current account of the state of the repository, so just running it at configure time is not satisfactory.  Putting a link to the .git directory is a disaster as git thinks you deleted every file in the repository.  I think a better approach would be to use Make functions (e.g. call) that would change directory into the GIT_BASE directory and run the git commands to generate the header file.  Setting the GIT_BASE variable in the generated makefile.  The makefile functions that implement the git queries could be stored in a configuration makefile that would be included in the build process.  For tarballs the configuration makefile could be replaced with one that set the function to copy the pre-generated include file instead.

**Comment by TimothyEDawson** [comment]
Good point regarding the symbolic link, at the very least it would ensure it is always detected as a "dirty" build. Tracking the root directory of the software in a variable like GIT_BASE and jumping over there for the git commands was what I originally planned to do.

I would like to use LOCI_SRC for that variable, as we use CHEM_SRC and GGFS_SRC in the same way, and the meaning seems clear to me. We had already used that variable in the CFDRC branch, although in that scenario it also replaced the build/OBJ directory, so we'll make sure to maintain the distinction going forward. LOCI_SRC will thus point to the root directory of the cloned Loci repo, and the only thing pointing to the build directory will be the Makefile generated by configure, no changes needed there at this time. Do you have any issues with that plan?

**Comment by EdwardALuke** [comment]
LOCI_SRC would work as well.  But I would like to see its use limited in the build process.   As I mentioned before, I will often compile sub-components of Loci (e.g. FVMtools, FVMMod, etc,) against a stable build of Loci.  For that I set the LOCI_BASE to the stable install (where no source code is available).  So for now I would like to see LOCI_SRC limited to the git version access so we don't end up loosing some of this capability in the future.

**Comment by TimothyEDawson** [comment]
Wanted to capture here some quick observations/ideas:

1. The output of `git describe --abbrev=6 --always --dirty --tags` is [most recent tag]-[number of commits since then]-g[first 6 characters of current commit hash, or more if not unique]["-dirty" if there are uncommitted changes]
2. We could set up the revision string to be "$Name: ${GIT_BRANCH}-${GIT_INFO} $", and then name the release branch "release" to produce a LOCI_REVISION_NAME formats like "release-v4.0.10", "dev-v4.1.b2", or "35-git-based-versioning-[...]-v4.1.b2-16-gcc369d-dirty".
3. It shouldn't be difficult to further preprocess the string to turn those into something like "rel-4.0.10", "dev-4.1.b2", and "issue35-4.1.b2-gcc369d".

**Comment by EdwardALuke** [comment]
Yes, this sounds like a good solution.  I don't think there is a real need to shorten release to rel or anything like that.  My naming of the releases was mainly constrained by the limitations of how cvs managed tags.  For example, dots could not be in CVS tags.    But the examples you give in item 2 would seem to be a pretty good solution.

-Ed

**Comment by TimothyEDawson** [comment]
@EdwardALuke When you get a moment, please take a look at this initial rough draft of the process and let me know your thoughts.

I spent too much time wrestling with tar. First I found the two super convenient options `--exclude-vcs` and `--exclude-vcs-ignores`. Then I found that not only are those options not available in all version of tar (some discussions mentioned it not being available on MacOS), but it doesn't even properly parse some of the directives in .gitignore (e.g. /Makefile should only ignore the top level file, but instead it was ignoring all Makefiles). Then right as I figured out the behemoth of a tar command I was happy with, I discovered the `git archive` command! It does not include any uncommitted changes when it generates an archive, but I would consider this to be desirable behavior. It prevents local changes from sneaking in, and one can always quickly commit a change locally in order to include it in an archive without needing to push it.

As mentioned in the commit message, I still need to get the revision string into the code itself. I also realized that I need to modify my rule for version.conf to force it to be regenerated on each call. I'll include those both in the next commit. I also just noticed some unnecessary escape characters in src/Makefile which I have removed.

**Comment by EdwardALuke** [comment]
It looks ok, but just a few comments.  I normally create a .tgz file instead of .tar which is a gzip compressed tar file.  Also, I don't think LOCI_SRC should be defined in the sys.conf file as this will be installed with the executable code install.  This generally isn't going to be around after the code is installed (it is not uncommon for me to build the code in a scratch directory and remove the untared directory after the code is installed).  So I think we need to work out a better way to manage the LOCI_SRC.  My suggestion is that we pass it in through the generated toplevel Makefile, e.g. 

tarball:
           LOCI_SRC=`pwd` @$(MAKE) -C OBJ $@

**Comment by TimothyEDawson** [comment]
I was considering compression, that's an option that I'll enable for sure.

Unfortunately LOCI_SRC must go somewhere inside the OBJ directory since it is needed to run the git commands for version.conf at compile time, and so it needs to be accessible when you're building from the OBJ directory. sys.conf seems to me the logical place to put it; however, perhaps it should be stripped out in the installation process so it doesn't make it to the install directory. That could be done by putting LOCI_SRC in yet another .conf file all its own which doesn't get installed at all, or by deleting that line from sys.conf as part of Install.bash.

**Comment by TimothyEDawson** [comment]
Alternatively, I suppose the $(OBJDIR)/version.conf file could contain something like
```
LOCI_SRC = /path/to/loci
GIT_INFO		=	"$(shell cd $(LOCI_SRC); git describe --abbrev=6 --always --dirty --tags)"
GIT_BRANCH		=	"$(shell cd $(LOCI_SRC); git rev-parse --abbrev-ref HEAD)"
```

And then when you run `make tarball` it stores a different file in src/version.conf which contains the results of the above commands - something like
```
GIT_INFO = v4.1.b3
GIT_BRANCH = beta
```

I think that will work, and that's more in line with your original suggestion. I'll test it out and see if that works in my typical build scenarios.

**Comment by EdwardALuke** [comment]
I think that would work.  I like that the LOCI_SRC is limited to the versioning support which I would prefer.

**Comment by TimothyEDawson** [comment]
Resolved with #42.
